ConditionalCompilation.md

```diff
@@ -1,28 +1,139 @@
+ * bodies of function declarations, switch statements, switch clauses, loop statements, for statements, while statements, if/else statements
+ * function return types
+ 
+> Note: The WGSL grammar currently does not allow attributes in front of const value declarations, variable declarations, directives, switch clauses and several statements. We would extend the syntax to allow them.
+> Which exacly we will enable is subject to discussion. Enabling them before all statements is perhaps undesirable.
+
+## `@if`, `@elif` and `@else` attributes
+
+Three new *compile-time attributes* are introduced.
+* An `@if` attribute takes a single parameter. It marks the decorated node for removal if the parameter evaluates to `false`.
+* An `@elif` attribute decorates the next sibling of a syntax node decorated by a `@if` or an `@elif`. It takes one parameter.
+   It marks the decorated node for removal if its parameter evaluates to `true` AND if the previous `@if` and `@elif` attribute parameters evaluate to false.
+* An `@else` attribute decorates the next sibling of a syntax node decorated by a `@if` or an `@elif`. It does not take any parameter.
+   It marks the decorated node for removal if the previous `@if` and `@elif` attribute parameters evaluate to false.
+
+*Example*
+
+```rs
+@if(feature_1 && (!feature_2 || feature_3))
+fn f() { ... }
+@elif(!feature_1)
+fn f() { ... }
+@else
+fn f() { ... }
+```
+
+## Execution of the conditional compilation phase
+
+The conditional compilation phase *should* be the first phase to run in the full WESL compilation pipeline.
+
+> Note: The conditional compilation was designed to be incremental. In case some features can only be resolved at runtime, the compiler can be invoked in two passes:
+> The first pass is invoked with compile-time features and returns a partially compiled WESL source. The second pass is invoked with the runtime features and returns a fully compiled WGSL source.
+
+1. The compiler is invoked with a list of features to *enable* or *disable*.
+2. The source file is parsed.
+3. The *compile-time features* in *Compile-time expressions* are resolved:
+    * If the feature is not present in the compiler's feature list, it is unchanged.
+    * If the feature is *enabled*, the identifier is replaced with `true`.
+    * If the feature is *disabled*, the identifier is replaced with `false`.
+   
+   The expression is *fully resolved* if all its *compile-time features* have been replaced by `true` or `false`.
+5. *Compile-time attributes* which have *fully resolved expressions* are evaluated:
+    * If the syntax node is marked for removal: it is eliminated from the source code along with the attribute.
+    * Otherwise, if all features in only the attribute is eliminated from the source code.
+4. The updated source is passed to the next compilation phase. (e.g. import resolution)
+
+## Possible future extensions
```

Comment:
Some old discussion suggested that default conditions ought to be set in the wgsl, for IDEs. WDYT? 

I suppose that could be handled subsequently to this PR, but I suggest adding it to the possible future section.

Reply:
I don't see an imediate need for this, and I don't see an elegant syntax for it, so I suggest we defer but keep in mind.